### DISCUSSION TITLE
Constructing 3-D Purkinje network and His bundle

### DISCUSSION THREAD

**Discussion by edorachlee** [rsachetto/MonoAlg3D_C#75] [General]
This time a discussion about creating my own Purkinje network.

I am seeing in [initialize_purkinje_with_custom_mesh](https://github.com/rsachetto/MonoAlg3D_C/blob/a7fc88a36dc27262b67eee97a878337416065896/src/purkinje_library/purkinje.c#L20) that we can read a Purkinje network file that is provided to us. However, all of the examples using _initialize_purkinje_with_custom_mesh_ seem to be simple 1-D examples, mostly using the simple_cable_2cm.vtk network file.

How would we approach constructing a full Purkinje network in a 3-D biventricular model, including a His bundle? Does that functionality exist?

**Comment by rsachetto** [comment]
You would have to construct your mesh, including the Purkinje network. The simulator only uses it as input. 

We have some papers about this:

https://www.nature.com/articles/s41598-023-38653-1
https://www.sciencedirect.com/science/article/pii/S1361841524000331

@bergolho (Lucas Berg) is a specialist at this subject. He implemented the Purkinje solver and added it in the simulator.

The software to generate the networks is open-source and available at:
https://github.com/bergolho/Shocker

**Comment by edorachlee** [comment]
Hi Rafael (and Lucas),

I have been able to generate some Purkinje networks using Shocker. However, I am currently running into "Duplicates are indexes" errors when conducting a simulation.

How should I fix the Purkinje network to get rid of the duplicates? I am able to provide a sample network file if necessary.

@bergolho your help would be appreciated also! Thank you.

**Reply by bergolho** [reply]
Hi @edorachlee,

I am glad to help you with the Purkinje networks! 

Could you please share with me the ".vtk" file you are using for the Purkinje together with the MonoAlg3D ".ini" configuration file. This way I can try to reproduce the error on my machine here and provide assistance on how to solve this issue and we can have a working simulation.

Best regards,
Lucas

**Reply by edorachlee** [reply]
[debug_config.zip](https://github.com/user-attachments/files/17680374/debug_config.zip)

Here is a debug environment! It is designed to be run in the MonoAlg3D folder. Thank you! BTW I am also Lucas.

Best,
Lucas

**Comment by bergolho** [comment]
Hi @edorachlee,

I managed to fix the errors on your Purkinje network and made two working examples in MonoAlg3D. The first one uses only the Purkinje network, which is normally how I debug the trees generated by the Shocker method in MonoAlg3D. The second one is with the Purkinje network coupled with the biventricular mesh you provided.

I noticed two minor problems with the Shocker network you sent me:

1) Initially, there was a cycle in a small set of branches located in the RV, by changing a little bit the position of the terminal that was linked to another branch I could fix this error. I believe this connection occurred during the post-processing step of the method when the non-connected PMJs were forced to connect. 

2) The other problem was related to the connection of points near 4 specific bifurcations within the network. The point before these bifurcations was erroneously linked to the point in front of this bifurcation. As a result, this will raise an error to the Depth-Search-First that I do over the Purkinje network to construct the mesh used by MonoAlg3D to solve the monodomain on the Purkinje. I had to correct these connections manually on the VTK. Because of that I am providing you with a new working network called _"merged_ps_fixed.vtk"_.

Later this week, I will revise the Shocker code and try to debug and fix these errors. There is a new Discussion tab on the Shocker repository that everyone can use now.

Also, there are some comments on the configuration files to help you execute the simulations with the Purkinje coupled to tissue.

Please let me know if you need any more help.

Best regards,
Lucas

[lucas_lee_monoalg3d_files.zip](https://github.com/user-attachments/files/17690967/lucas_lee_monoalg3d_files.zip)

**Reply by edorachlee** [reply]
Hi Lucas,

Thank you very much for looking into this. Your comments on the configuration files were really helpful! I am glad to have a working simulation and appreciate your help&insights.

Unfortunately, it seems that I have similar duplicate errors on any new Shocker networks I generate, so needless to say I am looking forward to having more discussions on the Shocker repository. I will see you there!

Best,
Lucas
